scionlab/models/user_as.py

```diff
@@ -192,117 +130,214 @@ class Meta:
-        self.attachment_point.trigger_deployment()
+        att_conf.link.update(active=att_conf.active)
+        att_conf.link.save()
+
+    def _delete_attachment(self, deleted_link: List[Link]):
+        """
+        Delete the `deleted_link` link
+        """
+        deleted_link.delete()
 
-    def is_use_vpn(self):
+    def _create_or_update_vpn_connection(self, att_conf) -> Tuple[Interface, Interface]:
         """
-        Is this UserAS currently configured with VPN?
+        Create or reuse a `VPNClient` with the specified `AttachmentPoint`.
+        The function also takes care of the attachment_point interface
+        :returns: public_ip and port of the `VPNClient`
         """
-        return VPNClient.objects.filter(host__AS=self, active=True).exists()
+        assert att_conf.attachment_point.vpn, "VPN not supported by this Attachment Point"
+        ap = att_conf.attachment_point
+        host = self.hosts.get()
+        # Reuse VPNClient if there's one with this AttachmentPoint
+        vpn_client = host.vpn_clients.filter(vpn=ap.vpn).first()
+        if not vpn_client:
+            vpn_client = ap.vpn.create_client(host, active=True)
+        elif not vpn_client.active:
+            vpn_client.active = True
+            vpn_client.save()
+
+        if att_conf.link is not None:
+            iface_ap, iface_client = att_conf.link.interfaceA, att_conf.link.interfaceB
+            iface_client.update(public_ip=vpn_client.ip,
+                                public_port=att_conf.public_port,
+                                bind_ip=None,
+                                bind_port=None)
+            iface_ap.update(
+                    border_router=ap.get_border_router_for_useras_interface(),
+                    public_ip=att_conf.attachment_point.vpn.server_vpn_ip
+            )
+        else:
+            border_router = BorderRouter.objects.first_or_create(host)
+            iface_client = Interface.objects.create(border_router=border_router,
+                                                    public_ip=vpn_client.ip,
+                                                    public_port=att_conf.public_port)
+            iface_ap = Interface.objects.create(
+                border_router=att_conf.attachment_point.get_border_router_for_useras_interface(),
+                public_ip=att_conf.attachment_point.vpn.server_vpn_ip)
+        return iface_client, iface_ap
 
-    def is_active(self):
+    def update(self,
+               label,
+               installation_type):
         """
-        Is this UserAS currently active?
+        Update this UserAS instance and immediately `save`.
+        Updates the related host, interface and link instances and will trigger
+        a configuration bump for the hosts of the affected attachment point(s).
         """
-        return self.interfaces.get().link().active
+        # XXX Useless since the model instance has already got the updated fields
```

Comment:
Actually that's exactly what the comment is saying, the `self.label = label` line is useless since `self.label` is already equal to `label` (and the same goes for the `installation_type`).